samples/core/Saving/Saving/Concurrency/Sample.cs

```diff
@@ -39,34 +41,34 @@ public static void Run()
                     {
                         if (entry.Entity is Person)
                         {
-                            // Using a NoTracking query means we get the entity but it is not tracked by the context
-                            // and will not be merged with existing entities in the context.
-                            var databaseEntity = context.People.AsNoTracking().Single(p => p.PersonId == ((Person)entry.Entity).PersonId);
-                            var databaseEntry = context.Entry(databaseEntity);
+                            var proposedValues = entry.CurrentValues;
+                            var databaseValues = entry.GetDatabaseValues();
 
-                            foreach (var property in entry.Metadata.GetProperties())
+                            foreach (var property in proposedValues.Properties)
                             {
-                                var proposedValue = entry.Property(property.Name).CurrentValue;
-                                var originalValue = entry.Property(property.Name).OriginalValue;
-                                var databaseValue = databaseEntry.Property(property.Name).CurrentValue;
+                                var proposedValue = proposedValues[property];
+                                var databaseValue = databaseValues[property];
 
-                                // TODO: Logic to decide which value should be written to database
-                                // entry.Property(property.Name).CurrentValue = <value to be saved>;
-
-                                // Update original values to 
-                                entry.Property(property.Name).OriginalValue = databaseEntry.Property(property.Name).CurrentValue;
+                                // TODO: decide which value should be written to database
+                                // proposedValues[property] = <value to be saved>;
                             }
+
+                            // Refresh original values to bypass next concurrency check
+                            entry.OriginalValues.SetValues(databaseValues);
                         }
                         else
                         {
-                            throw new NotSupportedException("Don't know how to handle concurrency conflicts for " + entry.Metadata.Name);
+                            throw new NotSupportedException(
+                                "Don't know how to handle concurrency conflicts for "
+                                + entry.Metadata.Name);
                         }
                     }
 
                     // Retry the save operation
                     context.SaveChanges();
```

Comment:
This example is flawed because another concurrency exception could happen here. When handling concurrency exceptions there usually needs to be a loop that executes the same SaveChanges in the same try/catch again and goes through resolution again until there is no concurrency exception or the save is aborted.

Reply:
Updated. Let me know if you prefer a different pattern. Maybe we should consider a new API for this :smile: